keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,914 @@
+in the kubelet configuration and eliminates the possibility of misaligned
+cgroup driver configuration between the kubelet and the runtime.
+
+## Motivation
+
+The responsibility of managing the Linux cgroups is currently split between the
+kubelet and the container runtime. Kubelet takes care of the pod (sandbox)
+level cgroups whereas the runtime is responsible for per-container cgroups.
+There currently are two different low-level management interfaces for cgroups:
+manipulating the cgroupfs directly or using the systemd system management
+daemon to manage them. Currently, both the kubelet and the container runtime
+has a configuration setting for selecting the cgroup driver (cgroupfs or
+systemd). These settings must be in sync, both kubelet and the runtime
+configured to use the same driver as the two drivers are incompatible because
+of a different kind of cgroups hierarchy used in them. Having kubelet and the
+container runtime to use non-matching cgroup drivers can cause hard to
+understand failures in container creation or inconsistent resource allocation
+on the node. This – two independent configuration settings for the same thing –
+is a common cause for user errors. Instead of having a split brain situation,
+there should be a single source of truth for the cgroup driver.
+
+### Goals
+
+- make kubelet automatically use the same cgroup driver as the container
+  runtime, making it unnecessary to specify cgroup driver in kubelet
+  configuration
+- maintain backward compatibility to run kubelet on top of an older container
+  runtime that doesn’t support this new feature
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I would like to simplify my node configuration by
+configuring fields just once, and not needing to synchronize between multiple
+processes.
+
+#### Story 2
+
+As a novice Kubernetes user, I would like to easily be able to start pods with
+all runtimes, even if they have differing opinions on defaults.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG Node and the feature will move along with CRI implementation
+adoption.
+
+## Design Details
+
+### CRI API
+
+Extend the CRI runtime API to inform the kubelet which cgroup driver should be
+used. A new RuntimeConfig rpc is added to query the information.
+
+```diff
+ // Runtime service defines the public APIs for remote container runtimes
+ service RuntimeService {
+ ...
++    // RuntimeConfig returns configuration information of the runtime.
++    rpc RuntimeConfig(RuntimeConfigRequest) returns (RuntimeConfigResponse) {}
```

Comment:
I don't think we need to determine this here, but here's something to bikeshed over: "runtime" is really overloaded. For instance, for @AkihiroSuda 's [RRO](https://github.com/kubernetes/enhancements/pull/3858) proposal, there could be a way for a runtime handler in a runtime implementation to specify runtime features for the oci runtime. 

I would propose we begin to use the phrasing `CRI implementation` and we could begin with `ImplementationConfig`. That opens opportunity for `RuntimeHandlerFeatures` to not conflict too much..

We can always debate this on the implementation PR though, I don't think we need to determine the name concretely before KEP freeze

Reply:
> I don't think we need to determine this here, but here's something to bikeshed over: "runtime" is really overloaded.

Yes, naming is always hard and I concur that it's not essential to try to finalize it here. We can bike-shed over it in the implementation and then update the KEP to reflect that. More important question is the structure of the API.